### PR TITLE
[test] Add new `MPI_Init_thread()` check

### PR DESCRIPTION
**Pull request by jgphpc**

### CODE REVIEW

File: cscs-checks/prgenv/mpi.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
+import reframe as rfm
+import reframe.utility.sanity as sn
+
+
+@rfm.required_version('>=2.14')
+@rfm.parameterized_test(['single'], ['funneled'], ['serialized'], ['multiple'])
+class Mpi_InitTest(rfm.RegressionTest):
+    def __init__(self, required_thread):
+        super().__init__()
+        self.descr = 'MPI_Init_thread ' + required_thread
```
Comment (teojgo) [inline cscs-checks/prgenv/mpi.py]: Delete the `self.descr` the default one generated by the parameterized test is ok.

File: cscs-checks/prgenv/mpi.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+        self.cppflags = {
+            'single':     ['-D_MPI_THREAD_SINGLE'],
+            'funneled':   ['-D_MPI_THREAD_FUNNELED'],
+            'serialized': ['-D_MPI_THREAD_SERIALIZED'],
+            'multiple':   ['-D_MPI_THREAD_MULTIPLE']
+        }
+        self.build_system.cppflags = self.cppflags[required_thread] + ['-static']
+        self.time_limit = (0, 1, 0)
+        found_mpithread = sn.extractsingle(
+            r'^mpi_thread_required=\w+\s+mpi_thread_supported=(?P<resA>\w+)\s+mpi_thread_queried=(?P<resB>\w+)\s+(?P<resC>\d)$',
```
Comment (teojgo) [inline cscs-checks/prgenv/mpi.py]: Split this string which is too long.

File: cscs-checks/prgenv/mpi.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+            'multiple':   2
+            # Output should look the same for every prgenv (cray, gnu, intel, pgi) / mpi_thread_multiple not supported:
+            # 'single':     ['mpi_thread_supported=MPI_THREAD_SINGLE mpi_thread_queried=MPI_THREAD_SINGLE 0'],
+            # 'funneled':   ['mpi_thread_supported=MPI_THREAD_FUNNELED mpi_thread_queried=MPI_THREAD_FUNNELED 1'],
+            # 'serialized': ['mpi_thread_supported=MPI_THREAD_SERIALIZED mpi_thread_queried=MPI_THREAD_SERIALIZED 2'],
+            # 'multiple':   ['mpi_thread_supported=MPI_THREAD_SERIALIZED mpi_thread_queried=MPI_THREAD_SERIALIZED 2']
+        }
+
+        self.sanity_patterns = sn.all([
+            sn.assert_found(r'tid=0 out of 1 from rank \s*(\d+) out of \s*(\d+)',
```
Comment (teojgo) [inline cscs-checks/prgenv/mpi.py]: Convert to 
```suggestion
            sn.assert_found(
                r'tid=0 out of 1 from rank \s*(\d+) out of \s*(\d+)',
                self.stdout),
```

File: cscs-checks/prgenv/mpi.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+                            self.stdout),
+            sn.assert_eq(found_mpithread, self.mpithread_version[required_thread])
```
Comment (teojgo) [inline cscs-checks/prgenv/mpi.py]: Move
```
 self.mpithread_version[required_thread]
```

To the next line

File: cscs-checks/prgenv/mpi.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+                                    'PrgEnv-intel', 'PrgEnv-pgi']
+        self.build_system = 'SingleSource'
+        self.sourcepath = 'mpi_init_thread.cpp'
+        self.cppflags = {
+            'single':     ['-D_MPI_THREAD_SINGLE'],
+            'funneled':   ['-D_MPI_THREAD_FUNNELED'],
+            'serialized': ['-D_MPI_THREAD_SERIALIZED'],
+            'multiple':   ['-D_MPI_THREAD_MULTIPLE']
+        }
+        self.build_system.cppflags = self.cppflags[required_thread] + ['-static']
```
Comment (teojgo) [inline cscs-checks/prgenv/mpi.py]: Split in two lines

File: cscs-checks/prgenv/mpi.py (hunk 1 above)
Comment (jgphpc) [reply cscs-checks/prgenv/mpi.py]: 3122756caa398ff5c756822db5e54276c0e8bae1

File: cscs-checks/prgenv/mpi.py (hunk 2 above)
Comment (jgphpc) [reply cscs-checks/prgenv/mpi.py]: 3122756caa398ff5c756822db5e54276c0e8bae1

File: cscs-checks/prgenv/mpi.py (hunk 3 above)
Comment (jgphpc) [reply cscs-checks/prgenv/mpi.py]: 3122756caa398ff5c756822db5e54276c0e8bae1

File: cscs-checks/prgenv/mpi.py (hunk 4 above)
Comment (jgphpc) [reply cscs-checks/prgenv/mpi.py]: 3122756caa398ff5c756822db5e54276c0e8bae1

File: cscs-checks/prgenv/mpi.py (hunk 5 above)
Comment (jgphpc) [reply cscs-checks/prgenv/mpi.py]: 3122756caa398ff5c756822db5e54276c0e8bae1

**Comment by jgphpc** [comment]
@jenkins-cscs retry all

**Comment by jgphpc** [comment]
```slurm_load_jobs error: Socket timed out on send/recv operation```
@jenkins-cscs retry all

**Comment by jgphpc** [comment]
@jenkins-cscs retry dom

File: cscs-checks/prgenv/mpi.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+        self.mpithread_version = {
+            'single':     0,
+            'funneled':   1,
+            'serialized': 2,
+            'multiple':   2
+# Output should look the same for every prgenv (cray, gnu, intel, pgi) / mpi_thread_multiple not supported:
+# 'single':     ['mpi_thread_supported=MPI_THREAD_SINGLE mpi_thread_queried=MPI_THREAD_SINGLE 0'],
+# 'funneled':   ['mpi_thread_supported=MPI_THREAD_FUNNELED mpi_thread_queried=MPI_THREAD_FUNNELED 1'],
+# 'serialized': ['mpi_thread_supported=MPI_THREAD_SERIALIZED mpi_thread_queried=MPI_THREAD_SERIALIZED 2'],
+# 'multiple':   ['mpi_thread_supported=MPI_THREAD_SERIALIZED mpi_thread_queried=MPI_THREAD_SERIALIZED 2']
```
Comment (vkarak) [inline cscs-checks/prgenv/mpi.py]: Can you remove these comments?

File: cscs-checks/prgenv/mpi.py [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+        }
+
+        self.sanity_patterns = sn.all([
+            sn.assert_found(
+                r'tid=0 out of 1 from rank \s*(\d+) out of \s*(\d+)',
```
Comment (vkarak) [inline cscs-checks/prgenv/mpi.py]: Why do you need such a complex regex with capturing groups, when you could achieve your purpose with a simple grep-like thing?

File: cscs-checks/prgenv/mpi.py [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+                                    'PrgEnv-intel', 'PrgEnv-pgi']
+        self.build_system = 'SingleSource'
+        self.sourcepath = 'mpi_init_thread.cpp'
+        self.cppflags = {
+            'single':     ['-D_MPI_THREAD_SINGLE'],
+            'funneled':   ['-D_MPI_THREAD_FUNNELED'],
+            'serialized': ['-D_MPI_THREAD_SERIALIZED'],
+            'multiple':   ['-D_MPI_THREAD_MULTIPLE']
+        }
+        #self.build_system.cppflags = self.cppflags[required_thread] + ['-static']
```
Comment (vkarak) [inline cscs-checks/prgenv/mpi.py]: Please remove this comment.

File: cscs-checks/prgenv/mpi.py [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+        self.build_system.cppflags  = self.cppflags[required_thread]
+        self.build_system.cppflags += ['-static']
+        self.time_limit = (0, 1, 0)
+        found_mpithread = sn.extractsingle(
+            r'^mpi_thread_required=\w+\s+mpi_thread_supported=(?P<resA>\w+)'
+            r'\s+mpi_thread_queried=(?P<resB>\w+)\s+(?P<resC>\d)$',
+            self.stdout, 3, int)
```
Comment (vkarak) [inline cscs-checks/prgenv/mpi.py]: Why do you use the `resA`, `resB` and `resC`, since you are only looking into the `resC`?

**Pull request merged**